### DISCUSSION TITLE
Artifact for community-dialects

### DISCUSSION THREAD

**Discussion by beikov** [hibernate/hibernate-orm#3783] [Design Proposals]
At the last PR-Meetup we found a few more PRs that want to add `Dialect`s for certain databases and we discussed the idea about creating a separate artifact/module that contains such community contributed `Dialect`s. This was also discussed on [Zulip](https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev/topic/Builtin.20Dialect.20requirements) where we were coming to a similar conclusion, that a separate artifact might be a good way forward.

So the main concerns for adding/keeping a `Dialect` to hibernate-core so far were:

* Who maintains it? Maintenance of the `Dialect` regarding bugs/features and `DialectChecks` for tests
* Who tests it? Run the testsuite to verify the `Dialect` works, ideally as Jenkins node connected to ci.hibernate.org

We maintain and test certain `Dialect`s for the big databases, so they are part of `hibernate-core`:

* H2
* Derby
* DB2
* MySQL
* Oracle
* PostgreSQL
* MariaDB
* SQL Server

Some `Dialect`s are part of `hibernate-core` because the vendors help with the maintenance and testing by e.g. providing a Jenkins node:

* SAP HANA
* Spanner
* Cockroach
* Sybase

We also have a few `Dialect`s for historic reasons that we don't want to remove from `hibernate-core` (yet?):

* HSQL
* Ingres
* Informix
* Mimer
* CUBRID
* Teradata
* MaxDB
* RDBMSOS2200
* TimesTen
* Cache
* Firebird

Here are the PRs for new `Dialects`:

* SQLite https://github.com/hibernate/hibernate-orm/pull/1874
* Impala https://github.com/hibernate/hibernate-orm/pull/3743
* InterSystems IRIS https://github.com/hibernate/hibernate-orm/pull/2335

A possible way forward for new `Dialect`s, and maybe also for some of the ones we currently have for historic reasons, could be to move them into a separate artifact e.g. `hibernate-community-dialects` that is part of the `hibernate-orm` Git repository. We would also do a community release for that, but we explicitly state that we do not provide support for these Dialects. These Dialects would be maintained on a best effort basis only. The community can provide PRs which the Hibernate team would consider, but the Hibernate team will not actively try to optimize such Dialects for performance or feature correctness.

The proposal is to do this for 6.0 and maybe also for 5.5 in case we decide to do that release.

What are your thoughts on this?

The proposal was accepted and will be implemented as part of https://hibernate.atlassian.net/browse/HHH-14693

**Comment by sebersole** [comment]
For completeness... another alternative is to simply reject such Dialect contributions that are not considered "important" or something we want to maintain.

**Comment by gbadner** [comment]
For the record, here is an article that shows EAP 7.x supported configurations: https://access.redhat.com/articles/2026253 .

Here are the tested DB configurations for EAP 7.3:

Database | JDBC Driver
-- | --
IBM DB2 e11.1 | IBM DB2 JDBC Universal Driver Architecture 4
Oracle 19c RAC | Oracle JDBC Driver v19
MySQL 8.0 | MySQL Connector/J 8
MariaDB 10.1 | MariaDB Connector/J 2
MariaDB Galera Cluster 10.1 | MariaDB Connector/J 2
Microsoft SQL Server 2017 | Microsoft SQL Server JDBC Driver 7
PostgreSQL 11.5 | JDBC4 Postgresql Driver, Version 42
Enterprise DB Postgres Plus Advanced Server 11.6 | Postgres Plus Advanced Server Driver 42
Sybase ASE 16 (SP02) | Sybase JDBC driver v16

RedHat tests all of the above DBs for Hibernate ORM versions that are included in RedHat products.

For sure, the dialects for the above DBs should remain in hibernate-core, or some other self-contained artifact that does not include dialects that are not tested.

I, personally, am in favor of moving all existing, non-tested dialects into a different artifact, such as `hibernate-community-dialects` that is part of the `hibernate-orm` Git repository. I suppose there could be business reasons for keeping particular non-tested dialects in hibernate-core. I will ask about that.

I'm not sure about doing this in Hibernate ORM 5.x though. Maybe starting in H6?

**Reply by gbadner** [reply]
To clarify, when I say "non-tested", I mean that it is not tested by either https://ci.hibernate.org/view/ORM/ or by QA on Hibernate ORM versions included in RedHat products.

**Reply by beikov** [reply]
>I suppose there could be business reasons for keeping particular non-tested dialects in hibernate-core. I will ask about that.

The move would obviously change the fully qualified class names of the Dialects, but we could still support automatic Dialect resolution for these databases.

The proposal is to start with this in Hibernate 6.0, maybe also in 5.5 if we decide to do that release.

**Comment by gavinking** [comment]
An opinion I hold, which I can’t perhaps justify very well, is that we should continue to support the classic “old” legacy enterprise RDBMs, even when we don’t have that many people using them, by which I mean Ingres, Informix, and Teradata.

I guess I feel like the sort of systems which are still using these old things are quite likely doing something rather important.

**Reply by gavinking** [reply]
MaxDB might also fit in that category; I’m not sure.

**Reply by beikov** [reply]
So am I understanding that right, that you are arguing that these Dialects which I categorized under "part of hibernate-core for historic reasons" should remain in hibernate-core? I don't have a strong opinion but it feels to me that these Dialects should move to `hibernate-community-dialects` unless their vendors step up and help with maintenance and testing. We currently have no way to do automated testing for most of these Dialects and since we do not support these Dialects through EAP either, I don't see how these are different from a quality perspective to community maintained Dialects.

**Reply by gavinking** [reply]
Well I'm not sure if you could characterize what I wrote as an "argument" exactly. More of a prejudice in favor of big old creaky systems that shift real money around.

> We currently have no way to do automated testing for most of these Dialects

Yes, sure, that is indeed a big problem.

**Comment by gavinking** [comment]
Does anyone know if CUBRID is widely used?

**Reply by beikov** [reply]
It seems it is still actively developed. Last release was in January 2021.

**Reply by gavinking** [reply]
Right, but do we have people using it with *Hibernate*, is the question.

**Reply by beikov** [reply]
Since I listed CUBRID in the "part of hibernate-core for historic reasons" list, I would treat it just like the others of that list. IMO we should just move them all to the `hibernate-community-dialects` artifact unless the vendor steps up. Even if the community uses these Dialects, IMO the differentiator for being in one or the other artifact should be the assured quality.

**Comment by dreab8** [comment]
I think the benefit of the new repository is to have a way to deal with new dialect implementations that is not just rejecting them. About the *historic* dialects I do not have a strong opinion, may be we can keep them all in core or decide to keep a subset of them and move the others to the `hibernate-community-dialects` repository but probably it is something we can decide later on.

**Reply by gavinking** [reply]
Well the other thing is that the current situation is that when we add new functionality it sometimes gets kinda haphazardly supported on some `Dialect`s but not on others with no really well-defined rule for which ones get it and which ones don't.

By moving some stuff outta there, we are much more likely to be able to exercise a bit of discipline, and make sure that new features get added to all the "team-supported" dialects (at least to those for which the functionality makes sense).

(I'm thinking especially about stuff related to DDL generation, and SQL functions, and features of HQL.)

**Reply by sebersole** [reply]
Wanted to make sure it was clear that the suggestion here is to move these to a different module within `hibernate-orm`, so kind of like `hibernate-envers`, `hibernate-jcache`, ...

I'm not understanding this comment within that idea.  Maybe you were thinking a repo separate from hibernate-orm altogether?

**Reply by sebersole** [reply]
Or maybe I'm confused ;)

But that was the proposal from the meeting - a separate module, not a new GitHub repo

**Reply by sebersole** [reply]
> By moving some stuff outta there, we are much more likely to be able to exercise a bit of discipline, and make sure that new features get added to all the "team-supported" dialects (at least to those for which the functionality makes sense).

This is the crux.  At the moment there is a perception that all of these Dialects are "blessed" which is not the case.  And that is what we want to get away from.  The Dialects we maintain and regularly test (or that are well maintained and tested by 3rd party) should be the only ones that are in hibernate-core imo.

I'm not against keeping some of the Dialects around that do not meet these criteria.  But I think there should be a compelling(-ish) reason to.

Of the specific ones you mention:

MaxDB:: That is an OSS database and we know is widely used with SAP.  And because it is OSS we can easily test it.  That is a compelling enough reason to keep it in hibernate-core imo.

Terradata:: Like HANA and Cache, there may be a business reason to keep this one around if you catch my drift

The others?  Meh.  Keep in mind also that this discussion is simply about better identifying which Dialects are blessed and which are just available in a "best effort" manner.  

And moving them to another module that gets published in no way stops users of that DB from using the Dialects; they'd simply have to depend on this other artifact, which is in fact exactly what we are trying to achieve - they have to go out of their way to accept that that particular Dialect is not belssed.

**Reply by dreab8** [reply]
sorry my bad, do not why I was thinking to another repository :( sorry again for the confusion :pray:

**Reply by beikov** [reply]
Hmm, the argument about it being open source and easily testable could apply to e.g. SQLite as well, yet we concluded that we don't want to support it and thus would only move it to `hibernate-community-dialects`. Like you wrote, these Dialects aren't tested and thus not "blessed" so I personally would rather see them moved. That way, we can really differentiate between "blessed"/tested and the best effort maintained ones.

**Reply by sebersole** [reply]
What is the "compelling" part of SQLLite?

**Reply by gavinking** [reply]
The "compelling" bit is that it's very popular. And so at first I was surprised you didn't want it. But now I have a hazy recollection that when I went and actually looked at it, it was a bit rubbish or a bit not-really-useful-from-Java or something and then I didn't want it much either.

But I don't recall very well.

**Reply by sebersole** [reply]
I have never used it, nor really know anything about it.  Andrea and Christian say it is used in highly embedded applications such as automotive and Android.  No one is using Hibernate in those spaces.

**Comment by mrotteveel** [comment]
From the perspective of such a 'lesser' dialect (Firebird), I wouldn't have objections to moving things to a community-dialects solution, or even having to provide our own library (e.g. a `org.firebirdsql:hibernate-dialect-firebird` library or something like that). Such separate libraries might also be a way to track actual usage (e.g. using Maven Central statistics)

However, I vaguely recall (a long time ago, maybe 10 years or so) that I had some issues when trying to create a custom dialect in its own package because dialects relied on some package-private classes or methods. I'm not sure if that recollection is accurate nor if it still applies today, but it would be necessary for dialects to be able to live in a different package (if only for the Java module system requirements of not having the same package in multiple modules).

**Reply by beikov** [reply]
That's not a problem anymore. Dialects can live in external artifacts, so there should be no issue. The benefit of you providing the Dialect is that you can choose to support any Hibernate version even if we decide not to maintain that version anymore. So we definitely encourage vendors to provide their own Dialect.

**Reply by gavinking** [reply]
>  I vaguely recall (a long time ago, maybe 10 years or so) that I had some issues when trying to create a custom dialect in its own package because dialects relied on some package-private classes or methods

If that's the case then we should totally fix that, because `Dialect` has always been a published extension point, and so if it is the case it's certainly unintentional.

**Reply by mrotteveel** [reply]
> That's not a problem anymore. Dialects can live in external artifacts, so there should be no issue. 

That is good to know, and to be clear, it could be that I'm misremembering things here.

> The benefit of you providing the Dialect is that you can choose to support any Hibernate version even if we decide not to maintain that version anymore. So we definitely encourage vendors to provide their own Dialect.

One thing to address would be things like https://github.com/hibernate/hibernate-orm/blob/wip/6.0/hibernate-core/src/main/java/org/hibernate/dialect/Database.java and replace or enhance that with a form of SPI / `java.util.ServiceLoader` form of discovery.

**Reply by beikov** [reply]
Yeah, I agree. I created https://hibernate.atlassian.net/browse/HHH-14478 for that.

**Comment by gavinking** [comment]
One issue with this is what happens with the code for guessing a dialect based on JDBC metadata. That's going to have to change somehow. The `Dialect`s are going to need to know how to guess themselves somehow.

**Reply by beikov** [reply]
See https://hibernate.atlassian.net/browse/HHH-14478

**Reply by sebersole** [reply]
This is trivial to do.

For completeness, it should also recognize JPA's settings such as `javax.persistence.database-product-name`, `javax.persistence.database-product-version`, etc

**Reply by sebersole** [reply]
https://github.com/hibernate/hibernate-orm/pull/3792
